### PR TITLE
[#10245][feat] AutoDeploy: Add Minimax M2 support

### LINKED ISSUE #10245: [Feature]:  Support Minimax model in AutoDeploy
### 🚀 The feature, motivation and pitch

https://huggingface.co/MiniMaxAI/MiniMax-M2/tree/main

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### LINKED ISSUE #10518: [Feature]:  AutoDeploy: support rms_norm for attention sharding heuristic
### 🚀 The feature, motivation and pitch

Minimax (and other models) may have rms norm inside the sharded attention region --> we must shard the rms norm weights like the rest of the weights

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by bmarimuthu-nv**
## Summary by CodeRabbit

* **New Features**
  * Added support for exporting MiniMax-M2 models with mixture-of-experts via PyTorch.
  * Added distributed RMSNorm operator for efficient sharded inference across multiple GPUs.
  * Enabled RMSNorm weight sharding in attention layers for improved distributed performance.

* **Tests**
  * Added validation tests for MiniMax-M2 mixture-of-experts patching.
  * Added multi-GPU tests for distributed RMSNorm operator.
  * Added transformation tests for RMSNorm sharding.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

## Description

Fixes [#10245](https://github.com/NVIDIA/TensorRT-LLM/issues/10245)

* Adds MoE patch for Minimax M2 MoE layer
* Fixes qk_norm weight sharding
  * q/k/v -> qk_norm -> attn -> o_proj. The qk_norm weights also need to be sharded on head dim same way as qkv.
* Adds some debug utils to dump graph IR after every transform based on the env var: `AD_DUMP_GRAPHS_DIR=<dir to dump graphs>`  

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py [hunk 1]
```diff
@@ -508,7 +508,7 @@ def flashinfer_mha_with_cache(
 
     # check if we need to re-combine outputs
     if num_prefill > 0 and num_decode > 0:
-        y = torch.empty_like(q)
+        y = torch.empty(q.shape, dtype=q.dtype, device=q.device)
```
Comment (bmarimuthu-nv) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py]: FakeTensor promises contiguous tensor. But the `q` can be non-contiguous (based on the model arch/presence of transpose) but end up contiguous at the end of the layer eventually. So we don't inherit the strides this way (and hence contiguous).

File: tensorrt_llm/_torch/auto_deploy/transform/library/sharding.py [hunk 2]
```diff
@@ -1863,41 +1863,150 @@ def _determine_fused_weight_dims(
...
 
-    slice_nodes = list(filtered_nodes(linear_node.users, ops=[torch.ops.aten.slice]))
-    if len(slice_nodes) > 0:
-        # we are probably in fused QKV case with single linear node and 3 slice nodes
-        assert len(linear_nodes) == 1
-        linear_node = linear_nodes[0]
-        assert all(
-            s.args[1] == 2 for s in filtered_nodes(linear_node.users, ops=torch.ops.aten.slice)
-        ), "Expecting slice nodes to slice tensor over dim=2"
-        fused_weight_dims = [s.args[3] - s.args[2] for s in linear_node.users]
```
Comment (bmarimuthu-nv) [inline tensorrt_llm/_torch/auto_deploy/transform/library/sharding.py]: Issue: slice ops are expected to be direct users of linear_node. But the `slice_nodes` obtained above can be anywhere downstream of linear.

Fix: Make the check strict - see if slice is directly after linear (as assumed by the logic here)

File: tensorrt_llm/_torch/auto_deploy/export/export.py [hunk 3]
```diff
@@ -218,6 +219,41 @@ def _clean_up_assertions_and_guards(gm: fx.GraphModule):
         canonicalize_graph(gm)
 
 
+def _rename_nodes_with_module_hierarchy(gm: fx.GraphModule) -> None:
```
Comment (bmarimuthu-nv) [inline tensorrt_llm/_torch/auto_deploy/export/export.py]: Util to make reading the graph/IR easy. Node names reflect the hierarchy they belong to. Example `model_layers_0_self_attn_q_norm_to_9` IR:

```
%model_layers_0_self_attn_q_norm_to_9 = aten.to.dtype(%model_layers_0_self_attn_q_proj_torch_linear_simple, torch.float32) : torch.float32
%model_layers_0_self_attn_q_norm_pow_2 = aten.pow.Tensor_Scalar(%model_layers_0_self_attn_q_norm_to_9, 2) : torch.float32
%model_layers_0_self_attn_q_norm_mean_1 = aten.mean.dim(%model_layers_0_self_attn_q_norm_pow_2, [-1], True) : torch.float32
%model_layers_0_self_attn_q_norm_add_2 = aten.add.Tensor(%model_layers_0_self_attn_q_norm_mean_1, 1e-06) : torch.float32
%model_layers_0_self_attn_q_norm_rsqrt_1 = aten.rsqrt.default(%model_layers_0_self_attn_q_norm_add_2) : torch.float32
%model_layers_0_self_attn_q_norm_mul_4 = aten.mul.Tensor(%model_layers_0_self_attn_q_norm_to_9, %model_layers_0_self_attn_q_norm_rsqrt_1) : torch.float32
```

File: tensorrt_llm/_torch/auto_deploy/utils/logger.py [hunk 4]
```diff
@@ -1,10 +1,102 @@
...
 
+    DUMP_GRAPHS_ENV = "AD_DUMP_GRAPHS_DIR"
+
+    def __init__(self):
+        super().__init__()
+        self._dump_dir = os.environ.get(self.DUMP_GRAPHS_ENV)
+        self._transform_counter = 0
+        self._dump_dir_initialized = False
+
+    def dump_graph(self, mod: nn.Module, transform_name: str, stage: str) -> None:
```
Comment (bmarimuthu-nv) [inline tensorrt_llm/_torch/auto_deploy/utils/logger.py]: Dump an SSA-style IR with Dtype annotation (default torch graph doesn't show dtype annotations)
Example:
```
%model_layers_0_self_attn_q_norm_to_9 = aten.to.dtype(%model_layers_0_self_attn_q_proj_torch_linear_simple, torch.float32) : torch.float32
%model_layers_0_self_attn_q_norm_pow_2 = aten.pow.Tensor_Scalar(%model_layers_0_self_attn_q_norm_to_9, 2) : torch.float32
%model_layers_0_self_attn_q_norm_mean_1 = aten.mean.dim(%model_layers_0_self_attn_q_norm_pow_2, [-1], True) : torch.float32
%model_layers_0_self_attn_q_norm_add_2 = aten.add.Tensor(%model_layers_0_self_attn_q_norm_mean_1, 1e-06) : torch.float32
%model_layers_0_self_attn_q_norm_rsqrt_1 = aten.rsqrt.default(%model_layers_0_self_attn_q_norm_add_2) : torch.float32
%model_layers_0_self_attn_q_norm_mul_4 = aten.mul.Tensor(%model_layers_0_self_attn_q_norm_to_9, %model_layers_0_self_attn_q_norm_rsqrt_1) : torch.float32
```

**Comment by bmarimuthu-nv** [comment]
@coderabbitai summary

File: tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py (hunk 1 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/custom_ops/flashinfer_attention.py]: flashinfer doesn't work with non-contiguous inputs. If `q` is non-contiguous at this point, we need ot make it contiguous before passing it into the kernels!

File: tensorrt_llm/_torch/auto_deploy/export/export.py (hunk 3 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/export/export.py]: do we think this could get too verbose? Otherwise, I think it's a great idea

File: tensorrt_llm/_torch/auto_deploy/transform/library/sharding.py [hunk 5]
```diff
@@ -1862,41 +1863,160 @@ def _determine_fused_weight_dims(
...
+            num_chunks = linear_chunk_users[0].args[1]
+            weight_dim = linear_node.meta["val"].shape[2]
+            fused_weight_dims = [weight_dim // num_chunks] * num_chunks
+
+
+def _shard_intermediate_attention_weights(
+    layer_subgraph: LayerSubgraph,
+    linear_nodes: List[Node],
+    transform_container: ShardingTransformContainer,
+) -> int:
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transform/library/sharding.py]: can we make this more conservative? Let's just shard weights associated with torch_rmsnorm

File: tensorrt_llm/_torch/auto_deploy/utils/logger.py [hunk 6]
```diff
@@ -1,10 +1,137 @@
...
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/utils/logger.py]: how is this different from this feature: https://github.com/NVIDIA/TensorRT-LLM/pull/10117/changes#r2676526174

**Review by lucaslie — APPROVED** [review]
